test/diag2dwd.jl

```diff
@@ -217,6 +218,317 @@ end
+  function compose_decapodes(decapodes_vars::Vector{Tuple{NamedDecapode{
+    Any, Any, Symbol}, Vector{Symbol}}}, relation::RelationDiagram)
+    # TODO: Calling this function "compose" is something of a misnomer.  e.g.
+    # You could very well give a vector containing a single decapode. Then the
+    # effect of applying the relation would be something like renaming
+    # variables. So this function should really be called oapply_decapodes (or
+    # just oapply if you are willing to import oapply from
+    # Catlab.WiringDiagrams).
+    r = relation
+    copies = decapodes_vars .|> first .|> copy
+    # TODO: We should also check that the number of variables given in the
+    # relation is the same as the number of symbols in the corresponding vector
+    # of Vars.
+
+    # Step -4: Determine the mapping of global ports to local ports.
+    # In a RelationDiagram, this is baked into the order of rows in the Port
+    # table.
+    # (i.e. for each (box,junction) pair, determine whether this corresponds to
+    # the 1st Var of that box, or the 2nd Var of that box, etc.)
+    # tuples (box, indices of rows with that box):
+    box_rows = [(b, incident(r, b, :box)) for b in unique(r[:box])]
+    local_ports = [map(y -> y-idxs[begin]+1, idxs) for (b, idxs) in box_rows]
+    # This is a column that one could hcat to the Ports table.
+    local_ports = local_ports |> flatten |> collect
+
+    # Step -3: Check that the types of all Vars connected by the same
+    # junction are the same.
+    foreach(parts(r, :Junction)) do j
+      # Check that all types are equal to the first type found.
+      j_idxs = incident(r, j, :junction)
+      first_box = copies[r[:box][j_idxs]] |> first
+      first_lp_idx = local_ports[j_idxs] |> first
+      first_type = first_box[:type][first_lp_idx]
+      foreach(r[:box][j_idxs], local_ports[j_idxs]) do b_idx, lp_idx
+        # Get the index of the row with this name in the box's Var table.
+        name = decapodes_vars[b_idx][2][lp_idx]
+        local_name_idx = incident(copies[b_idx], name, [:name]) |> only
+        # TODO: The `infer` type will never be here once the decapodes parsing
+        # is finished being refactored. So we don't check for it here.
+        # Note: This only returns the first type error found.
+        copies[b_idx][:type][local_name_idx] == first_type ||
+          error("The type of $(copies[b_idx][:name][local_name_idx]),
+            $(copies[b_idx][:type][local_name_idx]), in decapode
+            \"$(r[:name][b_idx])\" does not match the type of
+            $(first_box[:name][local_name_idx]), $(first_type), in decapode
+            $(r[:name][first(j_idxs)]). (Also, check that the order of the
+            decapodes you supplied matches the the order you specified in the
+            relation.)")
+      end
+    end
+
+    # Step -2: Do namespacing.
+    # Append each Var name with the name @relation gave the decapode.
+    for (copy_idx, box_name) in enumerate(r[:name])
```

Comment:
better to write 

```julia
for b in parts(r, :Box)
  boxname = r[b, :name]
  ...
```
That way it is clear what table you are iterating over. I have to check the schema to know that `:name` is an Attr of `:Box`